### ISSUE TITLE
[Muhammad] Prepare for team brainstorming of project architecture.

### ISSUE THREAD

**Issue by Funnear** [Funnear/agies#10]
**Process**:

1. Collaborators who wish to be involved in this, should clone this ticket and assign to themselves
2. We perform research to collect ideas individually, using own expertise, LLM prompting and so on. 
3. Propose time for brainstorming session, by which you will be ready.

**Topics for research**:

1) How **human-in-the-loop** expert systems are designed?
How do they differ from classic **Supervised Machine Learning**?
Note: We are not planning to invest in data labeling, we can avoid this if we think wisely.

- Best practices.
- Frameworks, libraries and other tooling recommendations.
- Examples of existing open-source projects.

2) How agentic systems are designed? 
Note: some agents will be MCP servers calling LLMs (we will start with just one), others will be for data pipeline, data analysis, ML, to add human-in-the-loop part, ontology and knowledge ghraphs.

- Best practices.
- Frameworks, libraries and other tooling recommendations.
- Examples of existing open-source projects.

3) How can we design the knowledge base?

- Can ontology be applied?
- Can knowledge graphs be applied?
- How can a workflow of updating the knowledge base can look like, if both - human domain experts and agents (ML and LLM) can contribute?
- How can knowledge be algorithmically extracted from this kind of knowledge base to be used in the agentic system workflow and/or data pipeline?

- Best practices.
- Frameworks, libraries and other tooling recommendations.
- Examples of existing open-source projects.

4) Which system design patterns are applicable?
Comparison of different patterns and recomendation what could work best for us, maybe in combination with each other.

------

_**Please, stick to these topics, do not spread your focus on audio data processing and so on. There will be place and time for everything in due course. You do not have to develop the entire project solo, we act as a team.**_

**Comment by mwaseem75** [comment]
I went through the four research areas. Below are my initial findings/recommendations for the project. I focused more on what could realistically work for our use case rather than listing too many tools.

## 1. Human-in-the-loop expert system

I think HITL should be used mainly for **uncertain/conflicting/novel cases**, rather than trying to label a large dataset manually.

The basic flow I see is:

Audio → ML/DL predictions → confidence/conflict check → automatic decision OR expert review → store expert decision/feedback

For example, if the models agree on Rock with high confidence, there is no reason to involve an expert. If one model says Rock and another says Metal, or the system detects a genre/subgenre which does not fit the current taxonomy, then we ask an expert.

This is different from classic supervised ML. In supervised ML, the main human effort is usually before/during training: humans create labelled data, the model is trained, and then predictions are mostly automated. In our case, we can start with existing datasets/models and use experts selectively during operation.

I think the human interaction should not be limited to "correct label". An expert could:

* approve/reject a prediction
* change the genre/subgenre
* provide a reason
* flag an unknown/new genre
* suggest a taxonomy change
* potentially add a rule or domain knowledge

The last points are interesting because the expert is contributing to the knowledge base, not just creating another training label.

Best practices I would suggest:

* Don't send every track to a human.
* Define clear conditions for human review: low confidence, model disagreement, taxonomy conflict, novel/outlier case, etc.
* Show the expert the evidence behind the prediction rather than only the final answer.
* Keep the original AI prediction even when the expert changes it.
* Record who made the decision, why, and which model/version produced the original prediction.
* Don't allow an LLM to directly make uncontrolled changes to the authoritative knowledge base.
* Allow "unknown / needs review" instead of forcing a classification.

For tooling, LangGraph looks interesting because it supports stateful workflows and human-in-the-loop interruptions. For the UI, something lightweight such as Streamlit should be enough for the hackathon.

Open-source projects/tools worth looking at:

* LangGraph - human-in-the-loop workflow patterns
* AutoGen - multi-agent/HITL patterns
* MTG-Jamendo dataset - useful existing music data rather than starting with our own labelling effort
* AcousticBrainz Genre Dataset - particularly interesting because of hierarchical/multi-label genre annotations

## 2. Agentic systems

My main takeaway here is that **not everything should be an LLM agent**.

I would divide the components into roughly three groups:

1. Deterministic/code based

   * metadata extraction
   * audio validation
   * dataset quality checks
   * taxonomy validation
   * workflow rules
   * confidence calculations

2. ML/DL based

   * spectrogram generation
   * audio feature extraction
   * audio embeddings
   * genre classification
   * similarity/outlier detection

3. LLM based

   * reasoning over conflicting evidence
   * semantic interpretation
   * explaining a classification
   * proposing new knowledge/taxonomy changes
   * interacting with the expert

So I would avoid creating an "agent" for every small operation. A useful principle could be:

**Agentize reasoning, not everything.**

For the architecture, I think a central orchestrator/supervisor makes sense. It can decide what needs to happen next based on the state of the current track.

Example:

ML confidence 95% + no conflicts → return result

ML confidence 55% → retrieve additional knowledge → LLM reasoning

Still uncertain / new genre → human review

This also helps with the project's goal of minimizing unnecessary LLM/token usage.

For the first LLM agent, I would keep its responsibility relatively narrow: give it the ML results, audio features, metadata and relevant knowledge and let it reason about the classification/conflicts.

MCP can be used as the interface between the LLM agent and tools/data. For example, the LLM could have tools such as:

* get genre information
* get subgenres
* search knowledge graph
* get track metadata
* get previous classification/review
* find similar tracks

MCP should not be considered the agent itself. It is more of a standardized way for the agent to access tools and external context.

Frameworks/tools I would investigate:

* LangGraph - workflow/orchestration + state + HITL
* MCP SDK - exposing tools/data to the LLM
* AutoGen - worth looking at for multi-agent collaboration, although I think LangGraph may fit our workflow better
* Python/PyTorch/sklearn for ML components
* Essentia/librosa/torchaudio for audio analysis

Open-source examples worth investigating include LangGraph, AutoGen, MCP reference implementations, and music/audio projects from the MTG ecosystem.

## 3. Knowledge base / ontology / knowledge graph

I think this could be one of the more interesting parts of the project.

I would separate three concepts:

**Knowledge Base** = the overall knowledge available to the system.

**Ontology** = defines the concepts and allowed relationships.

**Knowledge Graph** = stores actual entities and relationships using that structure.

For example, our ontology could define:

Entities:

* Track
* Artist
* Genre
* SubGenre
* AudioFeature
* Instrument
* Classification
* Expert
* Evidence

Relationships:

* SUBGENRE_OF
* RELATED_TO
* PERFORMED_BY
* HAS_FEATURE
* HAS_INSTRUMENT
* PREDICTED_AS
* CLASSIFIED_AS
* REVIEWED_BY
* SUPPORTED_BY
* CONTRADICTED_BY

Then the actual graph could contain something like:

Progressive Rock → SUBGENRE_OF → Rock

Track123 → PREDICTED_AS → Progressive Rock

Track123 → HAS_FEATURE → Complex Structure

Expert42 → REVIEWED → Track123

This also gives us provenance: we can keep track of where a classification or piece of knowledge came from.

I think a graph is particularly useful because music genres are not necessarily a simple tree. We can have parent/subgenre relationships, related genres, characteristics, instruments, similar tracks, expert decisions, evidence, etc.

For a hackathon, Neo4j looks like a pragmatic choice because it is easy to query and also gives us a good visual way to demonstrate the relationships.

I would not make the ontology too complicated initially. Start with a small useful model and extend it if needed.

### Knowledge update workflow

I would strongly recommend that agents do not directly modify the authoritative knowledge graph.

Something like:

Agent/ML/LLM proposes knowledge
↓
Validation
↓
Conflict/consistency check
↓
Human review (when required)
↓
Approve / modify / reject
↓
Versioned knowledge base

For example, an LLM could propose:

"Post-Metal should be added as a subgenre of Metal."

The proposal would contain evidence and then be reviewed before it becomes authoritative.

This prevents hallucinated LLM output from becoming permanent domain knowledge.

### How the knowledge can be used

The knowledge graph can be queried during classification.

For example:

Track → predicted genre → genre relationships → characteristics → related genres

The retrieved information can then be supplied to the reasoning agent.

We can also potentially combine the graph with vector search:

* vector search = "find tracks that sound/semantically resemble this track"
* knowledge graph = "what relationships and domain knowledge exist around these genres/tracks?"

Potential technologies:

* Neo4j
* RDF/OWL if we decide we need a formal ontology
* SHACL for validation/constraints
* Cypher/SPARQL for querying
* Qdrant/FAISS if we need vector similarity

For the MVP, I would probably start with Neo4j and only introduce RDF/OWL/SHACL or a separate vector DB if there is a clear need.

MusicBrainz and AcousticBrainz are also worth investigating as external sources of music/genre knowledge.

## 4. System design patterns

I looked at this more from the perspective of what we can actually build.

### A. Sequential pipeline

Audio → metadata → audio analysis → ML → LLM → human

Pros:

* very simple
* easy to implement

Cons:

* mostly fixed workflow
* LLM/human may be invoked even when not necessary

Good starting point, but probably too rigid for the final concept.

### B. Supervisor/orchestrator

A central orchestrator decides which component should run next.

Example:

```
             Orchestrator
            /      |       \
      Audio/ML   Knowledge   LLM
            \      |       /
             Decision Engine
                   |
             Human if needed
```

Pros:

* flexible
* works well with conditional workflows
* natural fit for HITL
* easy to control LLM usage

I think this is the strongest candidate for us.

### C. Blackboard/shared state

Each track gets a shared "case state" containing:

* metadata
* audio features
* ML predictions
* confidence
* retrieved knowledge
* LLM reasoning
* conflicts
* human decision

Different agents/components read from and contribute to this state.

This seems particularly suitable for our use case because the expert needs to see the complete evidence and the workflow may pause for human input and then continue.

I would combine this with the supervisor pattern rather than treating it as an alternative.

### D. Event-driven

TrackUploaded → AudioAnalyzed → PredictionCreated → ConflictDetected → HumanReviewRequested → ExpertDecision → KnowledgeUpdated

This is attractive for a production-scale system but probably more infrastructure than we need for a hackathon.

### My recommendation

I would combine:

**Supervisor/Orchestrator + shared case state + selective HITL + knowledge graph**

Something roughly like:

User uploads track
↓
Orchestrator
↓
Metadata + Audio/ML analysis
↓
Case State
↓
Decision/Confidence check
/             
confident          uncertain
↓                  ↓
result            Knowledge retrieval
↓
LLM reasoning
↓
still uncertain?
/       
no         yes
↓           ↓
result       human
↓
expert decision
↓
knowledge update

## Overall recommendation

If we want a practical MVP, I would start with:

* 1 orchestrator
* 1 LLM reasoning agent
* deterministic metadata/audio processing
* 1 or more existing audio/genre ML models
* a small genre ontology/knowledge graph
* selective human review
* provenance of predictions and expert decisions

Potential stack to investigate:

* LangGraph → orchestration/HITL
* MCP → tool access for the LLM
* Essentia/librosa/torchaudio → audio processing
* PyTorch/sklearn → ML
* Neo4j → knowledge graph
* Streamlit → expert UI
* Qdrant/FAISS → optional vector similarity

I would avoid starting with many agents, multiple LLMs, Kafka/Kubernetes, a very large ontology, or training a large model from scratch. We should first prove the end-to-end loop:

**audio → analysis → prediction → evidence/knowledge → reasoning → human review → knowledge update**

The main idea I see behind the project is:

**AI handles the volume, humans handle the ambiguity, and the knowledge base preserves what the system learns.**

**Comment by Funnear** [comment]
Thank you @mwaseem75 
I will copy this to our Miro board for the brainstorming session and share access with the team.
